src/core/or/circuitpadding_machines.c

```diff
@@ -0,0 +1,351 @@
+   * time frame, to obfuscate the fact that introduction circuits always send 4
+   * cells both ways */
+
+  /* Figure out the length of the BURST state so that it's randomized. See
+     definition of these constants for rationale. */
+  burst_state->length_dist.type = CIRCPAD_DIST_UNIFORM;
+  burst_state->length_dist.param1 = is_client ?
+    HS_MACHINE_PADDING_MINIMUM_CLIENT : HS_MACHINE_PADDING_MINIMUM_SERVICE;
+  burst_state->length_dist.param2 = is_client ?
+    HS_MACHINE_PADDING_MAXIMUM_CLIENT : HS_MACHINE_PADDING_MAXIMUM_SERVICE;
+
+  /* We mainly care about destroying circuit construction fingerprints. Here is
+   * an example intro circuit construction, where the timestamp granularity is
+   * 1 millisecond:
+   *
+   * IP circuit relay cells:
+   *  0: 11:56:59.703 -> EXTEND2
+   *  1: 11:57:00.074 <- EXTENDED2
+   *  2: 11:57:00.075 -> EXTEND2
+   *  3: 11:57:00.243 <- EXTENDED2
+   *  4: 11:57:17.406 -> EXTEND
+   *  5: 11:57:17.708 <- EXTENDED
+   *  6: 11:57:17.709 -> INTRODUCE1
+   *  7: 11:57:17.904 <- INTRODUCE_ACK
+   *
+   * So with about 200 milliseconds latency between each cell, we appropriately
+   * set the histogram edges to from 30 milliseconds to 100 milliseconds, to
+   * account for faster connections.
+   */
+
+  /* Histogram is: (30 msecs, 200 msecs, infinity). */
+  burst_state->histogram_len = 2;
+  burst_state->histogram_edges[0] = 30000;
+  burst_state->histogram_edges[1] = 200000;
+
+  /* Maximum number of tokens for the BURST state */
+  burst_state->histogram[0] = 1000;
+  burst_state->histogram_total_tokens = 1000;
+}
+
+/* Setup the BURST state of the machine that hides client-side rend
+ * circuits. If <b>is_client</b> is true, then this is the client-side part of
+ * this machine, otherwise it's the service-side of this machine. */
+static void
+setup_burst_state_for_hiding_rend_circuits(circpad_state_t *burst_state,
+                                           bool is_client)
+{
+  /* Token removal strategy for BURST state */
+  burst_state->token_removal =
+      CIRCPAD_TOKEN_REMOVAL_CLOSEST;
+
+  /* Histogram for BURST state:
+   *
+   * We want a machine that will send a bunch of fake cells over a reasonable
+   * time frame, to obfuscate the fact that introduction circuits always send 4
+   * cells both ways */
+
+  /* Figure out the length of the BURST state so that it's randomized. See
+     definition of these constants for rationale. */
+  burst_state->length_dist.type = CIRCPAD_DIST_UNIFORM;
+  burst_state->length_dist.param1 = is_client ?
+    HS_MACHINE_PADDING_MINIMUM_CLIENT : HS_MACHINE_PADDING_MINIMUM_SERVICE;
+  burst_state->length_dist.param2 = is_client ?
+    HS_MACHINE_PADDING_MAXIMUM_CLIENT : HS_MACHINE_PADDING_MAXIMUM_SERVICE;
+
+  /* Histogram is: (30 msecs, 200 msecs, infinity).
+   * See analysis above.
+  */
+  burst_state->histogram_len = 2;
+  burst_state->histogram_edges[0] = 30000;
+  burst_state->histogram_edges[1] = 200000;
+
```

Comment:
I think that doing the HTTP get thing here might help also.

And again we have to be careful about creating a thing that is now obviously unique. If we start padding before at least 3 hops are built, for example, then only padded intro/rend circs will be the ones with mismatched EXTEND2/EXTENDED2 request/response pattern. We can enforce the "don't pad until three hops" situation with a small histogram that just counts non-padding packets as our first state..

Reply:
Yes, that's a good idea! If we let intro circs build the circuit up to the intro point before padding, then the extra **EXTEND**->EXTENDED roundtrip (bold is outgoing) will blend in with the **BEGIN**->CONNECTED roundtrip that most normal circuits have (if you look at the backgammon board). We can start padding after that and then we will blend much better since then the **NEGOTIATE_PADDING**->INTRO1 will also blend with the **DATA**->DATA that most normal circuits have. 

Similarly for rend circuits, we should wait until the rendezvous circuit is established so that the **ESTABLISH_REND**->REND_ESTABLISHED dance, blends in with the **BEGIN**->CONNECTED dance of normal circuits.

Reply:
I have some good results by following this lead. I basically managed to hide the initial cell sequence of intro circs inside general circs by manipulating the allowed purposes: https://github.com/torproject/tor/pull/934/commits/90a81fca5736f165c9568375bf0205b5e9e9ca84#diff-c1ce03b78513b52ceac68cfe10c5f991R203

I tried to do the same for rendezvous circuits in that commit but it didn't work as well. It still stands out as pretty unique.

I'm attaching a backgammon board for the commit above when used on the real network. See how intro circuits are blending in with general circuits very well, but rendezvous circuits still stand out. I think I might be able to tweak the state machine transitions and perhaps get to where we want...

![Figure_1](https://user-images.githubusercontent.com/3611450/55973150-a0d57a80-5c8d-11e9-90c2-c6a7d304938a.png)